xarray/core/dataset.py

```diff
@@ -1077,12 +1077,41 @@ def relevant_keys(mapping):
         if not utils.is_scalar(dim) and not isinstance(dim, DataArray):
             dim = as_variable(dim, name='points')
 
+        indexers_dict = dict(indexers)
+        non_indexed = list(set(self.dims) - indexer_dims)
+
+        # need to generate a new dimension
+        points_len = lengths.pop()
+
+        variables = OrderedDict()
+
+        # add : slices for the non-indexed dimensions
+        slc = [indexers_dict[k] if k in indexers_dict else slice(None, None) for k in self.dims]
+        coords = [self.variables[k] for k in non_indexed]
+
+        # TODO need to figure out how to make sure we get the indexed vs non indexed dimensions in the right order
+        for name in data_vars:
+            var = self.variables[name]
+            if hasattr(var.data, 'vindex'):
+                variables[name] = DataArray(var.data.vindex[tuple(slc)],
+                                            coords=[np.arange(points_len)] + coords, dims=[dim] + non_indexed,
+                                            name=name)
+            else:
+                # TODO need to make sure this still works in the non-dask case
+                # if data is ndarray should give expected behaviour
+                variables[name] = DataArray(var.data[tuple(slc)],
+                          coords=[np.arange(points_len)] + coords, dims=[dim] + non_indexed,
+                          name=name)
+
+
+        return merge([v for k, v in variables.items() if k not in self.dims])
```

Comment:
I think you can just use the normal `Dataset` constructor instead of passing these into `merge`.